lxd/archive/archive.go

```diff
@@ -0,0 +1,71 @@
+package archive
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/lxc/lxd/lxd/apparmor"
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/shared/subprocess"
+)
+
+func ExtractWithFds(name string, args []string, stdin io.Reader, stdout io.Writer, s *state.State, profileName string) error {
```

Comment:
So I'm thinking this function signature can be improved for clarity and brevity:

1. Firstly it would be clearer if this function has a description comment so the reader can more easily understand what its intended purpose is without having to read the full implementation.
2. However from what I can tell, this function accepts a command `name` argument (would be clearer renamed to `command` or `cmd` in my view) to run with some `args` which it then runs connecting it with stdin and stdout.
3. However now it has apparmor profile support it also supports a `profileName` argument which is (without being stated) the expected output directory of the extraction (which presumably also needs to align with the `args` passed to the command).
4. So on the one hand this function seems to operate only on file handles, but the apparmor related arguments seem to indicate it is expected to write to file(s) inside a single directory.
5. I can't see any reason why we would need to accept stdout, as if we are expecting the command itself to write to a file path then we won't need to inspect stdout?

Would this work to be clearer?

```go
func ExtractWithFds(cmd string, args []string, stdin io.Reader, s *state.State, output os.File) error
```

By using an `os.File` as an argument it will force the understanding that we expect the output to be a file on disk and will allow the use of `output.Name()` for the apparmor profile. I think you should be able to use `f, err := os.OpenFile(dirPath, os.O_RDONLY, 0)` for directories.